jscomp/runtime/caml_obj.res

```diff
@@ -303,53 +299,46 @@ type eq = (Obj.t, Obj.t) => bool
     basic type is not the same, it will not equal 
 */
 let rec equal = (a: Obj.t, b: Obj.t): bool =>
-  /* front and formoest, we do not compare function values */
   if a === b {
     true
   } else {
     let a_type = Js.typeof(a)
-    if (
-      a_type == "string" ||
-        (a_type == "number" ||
-        (a_type == "bigint" ||
-          (a_type == "boolean" ||
-          (a_type == "undefined" || a === %raw(`null`)))))
-    ) {
+    if a_type !== "object" || a === %raw(`null`) {
       false
     } else {
       let b_type = Js.typeof(b)
-      if a_type == "function" || b_type == "function" {
-        raise(Invalid_argument("equal: functional value"))
-      } /* first, check using reference equality */
-      else if (
-        /* a_type = "object" || "symbol" */
-        b_type == "number" || (b_type == "bigint" || (b_type == "undefined" || b === %raw(`null`)))
-      ) {
+      if b_type !== "object" || b === %raw(`null`) {
         false
       } else {
         /* [a] [b] could not be null, so it can not raise */
         let tag_a = Obj.tag(a)
         let tag_b = Obj.tag(b)
-        if tag_a == 248 /* object/exception */ {
-          Obj.magic(Obj.field(a, 1)) === Obj.magic(Obj.field(b, 1))
-        } else if tag_a == 251 /* abstract_tag */ {
-          raise(Invalid_argument("equal: abstract value"))
-        } else if tag_a != tag_b {
+        if tag_a !== tag_b {
           false
-        } else {
+        } else if O.isArray(a) {
           let len_a = Obj.size(a)
           let len_b = Obj.size(b)
```

Comment:
Length check makes sense only for arrays, so moved it inside of the `if O.isArray(a)` block